### PR TITLE
First cut at bigtable support

### PR DESCRIPTION
**Pull request by jneem**
I've been poking at this for a while, and I think it's gotten to the point where it's worth asking whether I'm on the right track. Apart from wrapping the bigtable api, it does just enough of the bigtable table admin api to be able to run examples on the bigtable emulator.

The main missing things I know of are
-  [x] any testing whatsoever (including support for conveniently spinning up an emulator)
-  [x] retry handing for reading (but it's complicated enough that I think testing support is needed first)

### CODE REVIEW

File: src/bigtable/mod.rs [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+            inner: Box::pin(row_stream),
+        })
+    }
+
+    /// Queries a range of rows from a table, returning just the keys.
+    pub async fn read_row_range_keys(
+        &mut self,
+        table_name: &str,
+        range: impl RangeBounds<Bytes>,
+        rows_limit: i64,
```
Comment (rnarubin) [inline src/bigtable/mod.rs:313]: How do you feel about using `Option<u64>` for arguments like `rows_limit` and then converting to i64 (returning an error if the value exceeds the range)? Then `None` is more clear to say there's no limit than `0`

File: src/bigtable/mod.rs [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+    /// Set data for a collection of rows, all having the same table name and column family name.
+    ///
+    /// For more on the purpose of the `timestamp` field, see
+    /// [Google's docs](https://cloud.google.com/bigtable/docs/garbage-collection#timestamps).
+    pub async fn set_rows_data_with_timestamp<RowKey, RowData, ColName, CellData>(
+        &mut self,
+        table_name: &str,
+        family_name: String,
+        timestamp: i64,
```
Comment (rnarubin) [inline src/bigtable/mod.rs:482]: similar to `rows_limit`, what do you think about `timestamp: Option<u64>` which maps None to -1 internally?

File: src/bigtable/mutation.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+impl From<DeleteFromRow> for Mutation {
+    fn from(dfr: DeleteFromRow) -> Mutation {
+        Mutation::DeleteFromRow(dfr)
+    }
+}
+
+/// The error returned from a row mutation request.
+#[derive(Debug, thiserror::Error)]
+pub enum MutateRowsError {
```
Comment (rnarubin) [inline src/bigtable/mutation.rs:86]: consider adding `#[non_exhaustive]` to errors to keep them flexible without semver breaks

**Review by rnarubin — COMMENTED** [review]
This is really great work, thanks @jneem 

Some broad thoughts:
- I like the direction of the API, especially compared to the predecessor `solana-bigtable-connection` used in pixelstore. Streaming rows/bytes feels like the right abstraction rather than a whole vec at a time
- I'd prefer breaking changes like prost/tonic upgrades in separate PRs. It would help from a maintenance perspective to understand whether adding bigtable is a major or minor version change (ideally minor, but not a deal breaker). For the doc-comment build problem, prost-build 0.9 has a `disable_comments(proto_paths)` [option](https://docs.rs/prost-build/0.9.0/prost_build/struct.Config.html#method.disable_comments). This isn't a _great_ solution, but it's at least a reasonable workaround for those types that fail to build, and then a later PR can re-introduce the comments along with the version bumps

As an implementor of a new ya-gcp module, I'd also like to hear your opinions on the structure/patterns that the library has used. There are lots of improvements I'd like to make eventually, notably I think the Connector/DefaultConnector was the wrong layer to abstract; it would probably be better to abstract over the whole Service impl passed to tonic. Then users could do their own auth/load balancing/whatever, while the lib would provide a reasonable but replaceable default.

**Comment by jneem** [comment]
Thanks for the comments and sorry for the slow response! I've had a busy week, but I think I can get another version (without the breaking version bumps) ready soon after the Thanksgiving break.

File: src/bigtable/admin.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        /// Endpoint to connect to bigtable over.
+        @default("https://bigtableadmin.googleapis.com".into(), "BigtableTableAdminConfig::default_endpoint")
+        pub endpoint: String,
+    }
+}
+
+impl BigtableTableAdminConfig {
+    /// If the BIGTABLE_EMULATOR_HOST environment variable is set, use that variable to
+    /// override the endpoint.
+    pub fn with_emulator_env_var(mut self) -> Self {
```
[the remark this answers is not in the document]
Comment (jneem) [reply src/bigtable/admin.rs]: Fair enough, I've removed these pending a better handling for emulators. My original motivation, though, was the idea that application should basically always honor `BIGTABLE_EMULATOR_HOST` if it's set (AFAICT the go client library always does). So I was expecting to basically always call this function, and it's completely normal for the env var to be unset.

File: src/bigtable/mod.rs [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+        &mut self,
+        request: ReadRowsRequest,
+    ) -> Result<RowStream, tonic::Status> {
+        let response = self.inner.read_rows(request).await?.into_inner();
+
+        // Turn the response into a stream of rows.
+        // TODO: on errors, we should follow the retry policy. This is a little complicated, because
+        // on error we need to re-issue the query, while taking into account the last_scanned_row_key field
+        // to reduce re-querying. See the go implementation at
+        // https://github.com/googleapis/google-cloud-go/blob/21fc86d799bcdffc3db6f697b30ca48e072e7044/bigtable/bigtable.go#L171
```
[the remark this answers is not in the document]
Comment (jneem) [reply src/bigtable/mod.rs:248]: I'm not sure if I understood this correctly, but wouldn't non-smart re-queries result in the user seeing repeated rows in the stream? Anyway, it might be moot because the updated PR attempts the correct retry behavior.

**Comment by jneem** [comment]
Ok, so the updated PR reverts the various version bumps, using `disable_comments` to work around the failing doctests.

Regarding changes to the connection abstraction, it would be cool to decouple auth/load-balancing/etc. From the point of view of authoring a new module I don't think it makes much difference either way: all I did was copy a few `where` clauses from another module and everything else just worked. I guess that means it should be easy to adapt to whatever changes are made to the connection abstraction.

File: src/bigtable/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rnarubin) [reply src/bigtable/mod.rs]: Ah you're right, I hadn't considered elements already yielded to the caller. Anyway, yeah real retries are better

**Comment by jneem** [comment]
Ok, the testing isn't as comprehensive as it should be (especially regarding retries), but at least there's something there so I'll mark is as non-draft.

If you know of any easy way of injecting (retryable) errors into the gRPC stream, I'd very much like to know :)

**Comment by rnarubin** [comment]
Hi @k1nkreet. Yes fixing these tests is a great step.

1. Fixing emulators in CI would be a big help. Even the pubsub emulator doesn't work (not sure why). In the workflow there are skipped tests because they use pubsub but there was something broken about interacting with it. 
2. Upgrading to 1.61 should be fine, it's several releases old by now

**Comment by jneem** [comment]
Thanks for the ping! I made a separate PR for the MSRV bump. I'm less sure what to do about the emulator thing. I guess I'll try just updating the CI to include it, but if that doesn't work I might need some help...

File: src/bigtable/mod.rs [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+
+        // TODO: in the (presumably reasonably common case that the buffer comes
+        // in a single chunk, we should avoid copying the data.
+        self.value.extend_from_slice(&chunk.value);
+
+        if let Some(row_status) = chunk.row_status {
+            let row = self.finish_row();
+            if matches!(
+                row_status,
+                v2::read_rows_response::cell_chunk::RowStatus::CommitRow(_),
```
Comment (rnarubin) [inline src/bigtable/mod.rs:288]: There isn't a handler for the other RowStatus variant, `ResetRow` which requires the data for the row key to be discarded.

Given that functionality, it might be better to accumulate a `Vec<Bytes>` (or better yet `SmallVec<[Bytes; 4]>`) instead of copying bytes into a `BytesMut`. Then a reset doesn't incur wasted copying, and as a bonus the single-chunk case gets zero copies

File: src/bigtable/mod.rs [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+                    return;
+                }
+            } else {
+                return;
+            }
+
+            // If we get to this point, there was a retry-able error. We'll loop back and retry the
+            // request again, but first let's modify the request to avoid re-requesting previously-returned
+            // data.
+            if let Some(key) = last_key {
```
Comment (rnarubin) [inline src/bigtable/mod.rs:375]: there isn't a loop around the `inner.read_rows` call, this will exit the stream without retrying. There need to be 2 loops i think: 1 around inner.read_rows and 1 around response.next.

As a matter of convention, I prefer using explicit `continue`s instead of "fall-throughs" in these kinds of non-trivial repetitions (and in the case of nesting, labelled loops with labelled continues). That can more easily catch a case like this where it's hard to tell what portion of the logic needs to repeat. Here's an example from pubsub streaming https://github.com/standard-ai/ya-gcp/blob/d257f5684477bea2a59314e0515a20ce72c20a3f/src/pubsub/streaming_subscription.rs#L523-L538

**Review by rnarubin — COMMENTED** [review]
I remember reading (from another bigtable rust lib) about a conformance data set that google provides for testing edge cases in reads: https://github.com/googleapis/conformance-tests/blob/main/bigtable/v2/readrows.json https://github.com/liufuyang/bigtable_rs/issues/7#issue-1020095838

Adding these tests would help a lot with confidence, it's tricky state logic in read_rows. That doesn't necessarily have to be in this PR (maybe simple reads are good enough for users in feature store/pixel store?), but it should be written up as a task

File: src/bigtable/mod.rs [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+        true
+    }
+}
+
+impl RowSet {
+    /// Add a range of rows to this row set.
+    // TODO: it would be nice to take types more general than Bytes, but the
+    // RangeBounds trait makes that annoying: it doesn't allow extracting the
+    // start/end by value.
+    pub fn with_range(mut self, range: impl RangeBounds<Bytes>) -> Self {
```
[the remark this answers is not in the document]
Comment (rnarubin) [reply src/bigtable/mod.rs:114]: FWIW this is lower priority than MVP correctness. It doesn't have to happen in this PR, just an idea

File: src/bigtable/mod.rs (hunk 6 above)
Comment (jneem) [reply src/bigtable/mod.rs:288]: I think this is actually correct, although I agree that is isn't clear, and that `SmallVec<Bytes>` makes sense: `self.finish_row()` clears the current row, so if we get the `ResetRow`, we clear the current row from our state and return `None`.

File: src/bigtable/mod.rs [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+}
+
+impl RowRange {
+    // Take just the part of this range that's strictly after `key`, returning true if the resulting
+    // range is non-empty. (This API is a little weird; the point is that you're supposed to use
+    // it in `Vec::retain_mut`.)
+    fn restrict_to_after(&mut self, key: &Bytes) -> bool {
+        use v2::row_range::{EndKey, StartKey};
+        match &self.end_key {
+            Some(EndKey::EndKeyOpen(b)) | Some(EndKey::EndKeyClosed(b)) if b <= key => {
```
[the remark this answers is not in the document]
Comment (jneem) [reply src/bigtable/mod.rs:90]: I believe it is correct, and I'll add a comment explaining why

**Comment by jneem** [comment]
Thanks for the comments! I agree that the flushing is too permissive. If the bigtable server returns responses that violate the invariants, I guess the least bad option is to treat it as an unrecoverable error, right?

I will try to deal with the low-hanging fruit, and then I think maybe @k1nkreet will have some time to work on this? In any case I'll give him write access to the PR since I'll be off on Monday.

File: src/bigtable/mod.rs (hunk 7 above)
Comment (jneem) [reply src/bigtable/mod.rs:375]: Oh, good catch! As you can see, I haven't figured out a way to test the retry path. I thought about making a `tonic::transport::Channel` that injects errors somehow, but it didn't seem super easy. Do you know of a good way?

Comment (rnarubin) [reply src/bigtable/mod.rs]: I don't know an _easy_ way. The most robust way i can think of is to use the proto service schema to generate a stub bigtable server, and then mock messages+retriable errors. That's of course a lot of setup, but provides good abstraction without touching the client.

There could be easier solutions, I'm not sure

File: src/bigtable/mod.rs [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+struct ReadInProgress {
+    row: Row,
+    value: BytesMut,
+}
+
+impl ReadInProgress {
+    fn flush_bytes_in_progress(&mut self) {
+        if !self.value.is_empty() {
+            let value = self.value.split().freeze();
+            if let Some(cell) = self.last_cell_mut() {
```
[the remark this answers is not in the document]
Comment (matthew-healy) [reply src/bigtable/mod.rs:205]: Hey @rnarubin - I'm picking this up from @jneem to try & get it merged. 

As far as I can tell, what you're saying here is correct - `self.families` is only empty when the row is empty (in which case `self.value` should also be empty) ([source](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2#row)), and if we have a `family` then `family.columns` is never empty ([source](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2#google.bigtable.v2.Family)).

As a result, I'm now wondering what strength of assertion we need here. We could have each of these methods return a `Result` & update `BigtableClient`'s `read_rows` method to return `impl Stream<Item = Result<Row, SomeNewErrorType>>` - which is what I think you're proposing - but I'm not really sure how a client is meant to respond to an error that essentially says "we've ended up in an inconsistent state". Essentially, I'm wondering whether it might be valid to handle this with `panic!` or `unreachable!` (ideally combined with the conformance tests you mentioned above, which I'd like to work on), rather than bubbling it up to the consumer. What do you think?

[the remark this answers is not in the document]
Comment (rnarubin) [reply src/bigtable/mod.rs:205]: Yes i think an error is the right call instead of a panic, though acknowledging that this error isn't really actionable. I'd consider panics as conventionally "our software reached an impossible state/did something wrong", but here the cause would be a misbehaving server -- or more generally, bad network inputs. An error feels more appropriate when the cause is out of our control.

If the user really wants, they could conceivably retry. A panic is more terminal

**Comment by xlambein** [comment]
Using this PR yesterday, I was in a situation yesterday where I had to call `BigtableClient::mutate_rows` directly, rather than go through e.g. `set_rows_data`.  Doing this, I got a bigtable error saying the table name didn't exist, which I eventually found out was because methods like `set_rows_data` prepend a required prefix to the name:

https://github.com/standard-ai/ya-gcp/blob/b2e3ea7b0661c173fa20d521ec9f096ea939fb3d/src/bigtable/mod.rs#L554

IMO either this should be documented in `mutate_row(s)`, or these lower-level methods should also prepend the prefix.  In the case of the former, `BigtableClient` should have a method to construct a fully-qualified table name, because currently the prefix is private.

**Comment by matthew-healy** [comment]
@xlambein - Perhaps @jneem can correct me here, but it seems like the design intention was for `mutate_rows` to be 1:1 with the `MutateRows` rpc (which you can find documented [here](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2)). As such it directly uses the proto-generated `MutateRowsRequest` type as its argument. That type has a `table_name` field which is expected to include the fully-qualified table name. 

My initial feeling is that it'd be incongruous to still take that type as an argument but then opaquely change the `table_name`. Equally, the current state of affairs will clearly lead to mistakes.

I think the best way forward is to change the `mutate_rows` and `mutate_row` methods to no longer directly take the proto types as arguments, and then build the proto requests privately so that any call to the `BigtableClient` is always building the `table_name` from the prefix & supplied name. 

That said, I'm interested to hear what @jneem and @rnarubin think of this.

**Comment by matthew-healy** [comment]
Also @rnarubin - I think I've resolved all of the blocking comments, so have re-requested you review.

**Comment by rnarubin** [comment]
Re: mutate_rows vs set_row*

I agree with having a low-level function that matches the rpc API as closely as possible. Because there's a lot of fine-grained behavior defined in the message request, replicating that control in-library and then constructing the request internally would be difficult to maintain (probably lots of boilerplate, subtle behaviors). So i would advise keeping the low-level one with just the generated proto message request as an argument, for users who need that control.

To address @xlambein 's problem, adding a method to construct the fully specified table name is a good idea. It's a shame that the bigtable schema doesn't document the name format on the field in `MutateRowsRequest`, they _do_ in the `MutateRowRequest` (note singular)

    /// Required. The unique name of the table to which the mutations should be applied.
vs

    /// Required. The unique name of the table to which the mutation should be applied.
    /// Values are of the form
    /// `projects/<project>/instances/<instance>/tables/<table>`.

**Comment by jneem** [comment]
I realized that I was wrong about the truncation of open intervals, so I pushed a fix.

**Review by rnarubin — COMMENTED** [review]
A question @matthew-healy / @jneem / @xlambein -- how valuable is landing this PR as an MVP compared to (what sounds like the existing workflow) importing this code as a git reference.

The code here is good work, and I think it's shaping up well. However there's a lot of nuanced behavior in the BigTable service that I would not consider this production ready without more testing code:
- the conformance testing i mentioned above
- some kind of testing of retries (resuming at the right keys is a big area, I'd like a test for that truncation bug!)

If this is valuable as an unblocker, I am open to landing this PR and deferring tests given a few additions:
1. a fix on the constructed RowSet on retry mentioned below
2. a note in the top-level readme that BigTable support is still alpha, similar to GCS
3. links to filed tasks that include the testing initiatives

**Comment by matthew-healy** [comment]
@rnarubin It's definitely valuable to have this merged, but as it's not currently a blocker for us I'm not sure it's necessary to rush it. I'm going to try and implement the conformance tests this week, and then I'll take a look at setting up a stubbed Bigtable server for the other test cases. 

If, for whatever reason, this becomes a blocker, then we can reconsider.

**Comment by matthew-healy** [comment]
@rnarubin Hey! So I spent a bit of time last week working on the conformance tests. It took a little longer than I was hoping, but I managed to get them running. Unfortunately - as you predicted - they don't pass. In fact, the very first one fails because we seem to ignore the `commitRow` field in the `CellChunk` message. (I'm also having a bit of a hard time getting everything to implement `UnwindSafe` so I can catch the panics, so I don't yet have a good picture of how many of the tests fail, but I do know it's at least not all of them :sweat_smile: ).

That being said, I think the issues we have are mainly around poorly-behaved servers, as we're successfully using the branch without issues right now. Given the investment required to get this past alpha-maturity, I think it might be best if we do what you suggested & clean up what we have so we can merge this PR, and I'll open tickets for the remainder of the testing work.

I'll get started on the tasks you mentioned now and ping you for another review when they're done. I'll continue working on the conformance tests in-between other tasks and hopefully follow up with a PR in a few weeks.

**Comment by matthew-healy** [comment]
I've added Bigtable support (& its alpha status) to the README, and fixed the retry issue identified above. I also created #18 and #19 to cover the requested test cases.

I'm going to tidy up my (very rough) WIP branch & open a draft PR with the conformance tests to increase the bus factor here a little too. :bus:

**Review by rnarubin — APPROVED** [review]
Sounds like a plan :+1:
